File: py/entities/bar.py
"""An [open, high, low, close, volume] bar."""

from datetime import datetime


class Bar:
    """Represents an OHLCV price bar."""

    def __init__(self, time: datetime, open: float, high: float, low: float, \
                 close: float, volume: float):
        self.time = time
        self.open = open
        self.high = high
        self.low = low
        self.close = close
        self.volume = volume

    def is_rising(self) -> bool:
        """Indicates whether this is a rising bar (open < close)."""
        return self.open < self.close

    def is_falling(self) -> bool:
        """Indicates whether this is a falling bar (close < open)."""
        return self.close < self.open

    def median(self) -> float:
        """The median price: (low + high) / 2."""
        return (self.low + self.high) / 2

    def typical(self) -> float:
        """The typical price: (low + high + close) / 3."""
        return (self.low + self.high + self.close) / 3

    def weighted(self) -> float:
        """The weighted price: (low + high + 2*close) / 4."""
        return (self.low + self.high + self.close + self.close) / 4

    def average(self) -> float:
        """The average price: (low + high + open + close) / 4."""
        return (self.low + self.high + self.open + self.close) / 4

    def __repr__(self) -> str:
        return f"Bar({self.time}, {self.open}, {self.high}, {self.low}, " \
               f"{self.close}, {self.volume})"

File: ts/entities/bar.ts
/** An [open, high, low, close, volume] bar. */
export class Bar {
  /** The date and time.
   *
   * For _ohlcv_ bar entities it corresponds to the closing time, so that an _ohlcv_ bar accumulates lower-level entities
   * up to the closing date and time.
   */
  time!: Date;

  /** The opening price. */
  open!: number;

  /** The highest price. */
  high!: number;

  /** The lowest price. */
  low!: number;

  /** The closing price. */
  close!: number;

  /** The volume. */
  volume!: number;

  constructor(data?: any) {
    if (data) {
      for (const property in data) {
        if (Object.prototype.hasOwnProperty.call(data, property)) {
          (this as any)[property] = data[property];
        }
      }
    }
  }

  /** Indicates whether this is a rising bar (open < close). */
  isRising(): boolean {
    return this.open < this.close;
  }

  /** Indicates whether this is a falling bar (close < open). */
  isFalling(): boolean {
    return this.close < this.open;
  }

  /** The median price, calculated as _(low + high) / 2_. */
  median(): number {
    return (this.low + this.high) / 2;
  }

  /** The typical price, calculated as _(low + high + close) / 3_. */
  typical(): number {
    return (this.low + this.high + this.close) / 3;
  }

  /** The weighted price, calculated as _(low + high + 2*close) / 4_. */
  weighted(): number {
    return (this.low + this.high + this.close + this.close) / 4;
  }

  /** The average price, calculated as _(low + high + open + close) / 4_. */
  average(): number {
    return (this.low + this.high + this.open + this.close) / 4;
  }
}
